### ISSUE TITLE
RIP: Web3 frontend Library

### ISSUE THREAD

**Issue by midgerate** [raid-guild/RIPs#78]
# Project Idea

A react.js + hooks library that makes connecting and interacting with smart contracts very easy. 
https://github.com/raid-guild/quiver/tree/main/web3

### Project Submitted By

Midgerate

## Summary

**Quiver aims**
- to decrease the time it takes to integrate from 30 hours to 15 minutes
- to make it faster to add more contract interactions to their codebase
- to focus on the business logic rather than web3 specifics.

## Why should we build this?

1. It's a very time-consuming process to integrate a smart contract with a react web app. (Create a context, expose methods to connect with the wallet, read from and write to the contract, parse and convert tokens and errors, and so on.)
2. Every developer does it in their own style, which makes it harder for other devs to understand.

## What's done?

Most of the hooks and helpers are already created and are available for the same at https://github.com/raid-guild/quiver/tree/main/web3

## What is left?

1. Convert quiver into an npm library that also installs the dependencies with it. 
2. Expose all configuration options to the user of the library. 
3. Add subgraph hooks and utilities. 
4. Add other useful hooks that I haven't personally got exposed to. 

## Raid Party Skills Needed
React, Ethers, Typescript

## Cost (in USD)

TBD

The payment can be done in DAI / RAID etc based on the split that can be decided later.

**Comment by xivanc** [comment]
@midgerate regarding `Add subgraph hooks and utilities.` , will it have the ability to fetch different graphQL servers in one project?

**Comment by xivanc** [comment]
as well wouldn't it make sense to isolate `3. Add subgraph hooks and utilities.` part in a separate npm package? to keep our dependencies small as not every project necessarily needs both web3 and GQL

**Comment by midgerate** [comment]
@xivanc 

Just like you can use multiple contracts like 
```
const {contract: contractA} = useContract(contractAAddress, ContractAFactory)
const {contract: contractB} = useContract(contractBAddress, ContractBFactory)
```

You would be able to access multiple subgraphs on different networks and hosts at the same time. 

The subgraph might / might not use apollo client or something similar (I haven't decided on it yet) 
Depending on that we can decide if including a dependency is going to take a hit on the users or not. 

We could make the client a dynamic import and make the client as a peer dependency which will ensure it is not loaded in the library by default. 

Long story short, there are many things we can consider here, which is why I made it a RIP.

**Comment by xivanc** [comment]
> You would be able to access multiple subgraphs on different networks and hosts at the same time.

maybe i would consider to use more general naming here as GQL fetching don't need to be always connected to The Graph protocol data fetching, but in general we could query any (no only The Graph protocol) endpoint. Or do we want to stick only to The Graph protocol context here?

**Comment by plor** [comment]
Would it be possible to put this RIP as a `.md` document in the proposals directory of this repo? It seems we have neglected this part of the process lately, but it is important to have a reference document referring to the initiatives that went through the RIP process and being able to refer to it as such. I think we can "close" this issue at that point.
